### PR TITLE
Update to allow _revinclude=*:* search and remove ResourceType expansion for Global Search

### PR DESCRIPTION
**Pull request by feordin**
## Description
This updates the search service and search implementations to allow for _revinclude=*:* searches.
For Global searches, Resource Type Expansion and IsHistory removed from Where clause.

## Related issues
Addresses [issue AB#105303].

## Testing
A new E2E test was added, and manual search .http file was also updated

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

**Comment by SergeyGaluzo** [comment]
> 

Please add IF EXISTS (SELECT * FROM sys.indexes WHERE object_id = object_id('ReferenceSearchParam') AND name = 'xxxx') check

---
Refers to: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/68.diff.sql:12 in 668292f. [](commit_id = 668292fd8db41c21a0070f057948e4630ae86059, deletion_comment = False)

**Review by SergeyGaluzo — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql [hunk 1]
```diff
@@ -42,3 +42,15 @@ WHERE IsHistory = 0
 WITH (DATA_COMPRESSION = PAGE)
 ON PartitionScheme_ResourceTypeId(ResourceTypeId)
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql:43]: Please remove this index

File: src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj [hunk 2]
```diff
@@ -75,6 +75,9 @@
     <Compile Remove="**\*.generated.*.cs" />
     <Compile Include="**\*.generated.$(TargetFramework).cs" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="Features\Schema\Migrations\68.diff.sql" />
+  </ItemGroup>
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj:81]: Please remove.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql (hunk 1 above)
Comment (feordin) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql:43]: I removed the index.  Should I also put a DROP INDEX statement in the 68.diff.sql ? To remove the existing index on upgrade?  Or should we leave it there?

File: src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj (hunk 2 above)
Comment (feordin) [reply src/Microsoft.Health.Fhir.SqlServer/Microsoft.Health.Fhir.SqlServer.csproj:81]: Removed.

**Comment by feordin** [comment]
> > 
> 
> Please add IF EXISTS (SELECT * FROM sys.indexes WHERE object_id = object_id('ReferenceSearchParam') AND name = 'xxxx') check
> 
> Refers to: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/68.diff.sql:12 in [668292f](https://github.com/microsoft/fhir-server/commit/668292fd8db41c21a0070f057948e4630ae86059). [](commit_id = [668292f](https://github.com/microsoft/fhir-server/commit/668292fd8db41c21a0070f057948e4630ae86059), deletion_comment = False)

IF EXISTS added.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql (hunk 1 above)
Comment (SergeyGaluzo) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql]: Please DROP

**Comment by SergeyGaluzo** [comment]
> 

FYI BEGIN/END are not required here.

---
Refers to: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/69.diff.sql:21 in 93b6794. [](commit_id = 93b6794feb12a2c140397dc4e98035d77f196556, deletion_comment = False)

**Review by SergeyGaluzo — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql [hunk 3]
```diff
@@ -24,21 +24,15 @@ ON dbo.ReferenceSearchParam
...
+CREATE UNIQUE INDEX IXU_ReferenceResourceId_ReferenceResourceTypeId_SearchParamId_BaseUri_ResourceSurrogateId_ResourceTypeId ON dbo.ReferenceSearchParam 
+  ( 
+    ReferenceResourceId
+   ,ReferenceResourceTypeId
+   ,SearchParamId
+   ,BaseUri
+   ,ResourceSurrogateId
+   ,ResourceTypeId
+  )
+  WITH (DATA_COMPRESSION = PAGE, ONLINE = ON)
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Tables/ReferenceSearchParam.sql:36]: >, ONLINE = ON) [](http://example.com/codeflow?start=31&length=14)

Please remove ONLINE = ON from here.

**Comment by SergeyGaluzo** [comment]
I intentionally asked for index not filtered on IsHistory = 0, because our end goal is to remove IsHistory from search param tables. If we still continue adding IsHistory = 0 in generated queries it will cause extra key lookup. Can you remove IsHistory = 0 when accessing TokenSearchParam?

**Review by SergeyGaluzo — CHANGES_REQUESTED** [review]
:clock1:

**Review by SergeyGaluzo — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/69.diff.sql [hunk 4]
```diff
@@ -0,0 +1,20 @@
+
+IF EXISTS (SELECT * FROM sys.indexes WHERE object_id = object_id('ReferenceSearchParam') AND name = 'IXU_ReferenceResourceId_ReferenceResourceTypeId_SearchParamId_BaseUri_ResourceSurrogateId_ResourceTypeId')
```
Comment (rajithaalurims) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/69.diff.sql:2]: This should be if Not Exists then create correct?

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/69.diff.sql [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+BEGIN
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/69.diff.sql:3]: BEGIN/END are not needed.

**Review by SergeyGaluzo — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs [hunk 6]
```diff
@@ -497,7 +497,10 @@ private void HandleTableKindNormal(SearchParamTableExpression searchParamTableEx
 
             using (var delimited = StringBuilder.BeginDelimitedWhereClause())
             {
-                AppendHistoryClause(delimited);
+                if (HasTypeReference(searchParamTableExpression))
+                {
+                    AppendHistoryClause(delimited);
+                }
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:503]: This should not be limited to refs.

**Comment by SergeyGaluzo** [comment]
>     internal class SqlQueryGenerator : DefaultSqlExpressionVisitor<SearchOptions, object>

Where are we removing resource type expansion?

---
Refers to: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs:28 in 2593844. [](commit_id = 25938449e9a9c5152147e4b9b27518c5a76857bf, deletion_comment = False)

**Pull request closed without merging**